### PR TITLE
added the Adam method

### PR DESCRIPTION
**Pull request by yorkerlin**
@karlnapf 
I will add the doc soon.

### CODE REVIEW

**Comment by yorkerlin** [comment]
@karlnapf 
I think it will be great that a contributor can add the `gist` link about how to generate reference results.

The gist link for this PR is
https://gist.github.com/yorkerlin/6ed61e80557fd94d10fc533f901f99c8

**Comment by karlnapf** [comment]
Very good idea with the gist link actually
Please dont merge code without doc

File: src/shogun/optimization/AdamUpdater.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+{
+	REQUIRE(epsilon>0,"Epsilon (%f) must be non-negative\n",
+		epsilon);
+	m_epsilon=epsilon;
+}
+
+void AdamUpdater::set_first_moment_decay_factor(float64_t decay_factor)
+{
+	REQUIRE(decay_factor>0.0 && decay_factor<=1.0,
+		"decay factor (%f) for first moment must in (0,1]\n",
```
Comment (karlnapf) [inline src/shogun/optimization/AdamUpdater.cpp]: Decay (capital D)

File: tests/unit/optimization/StochasticMinimizers_unittest.cc [hunk 2]
```diff
@@ -946,6 +947,39 @@ TEST(AdaDeltaUpdater, test1)
...
+	//cost=
+	//0.368617
+	float64_t cost=bb->get_cost()/(float64_t)bb->get_sample_size();
+	EXPECT_NEAR(cost,0.368617,1e-5);
+	SGVector<float64_t> w=bb->obtain_variable_reference();
+	EXPECT_NEAR(w[0],1.49674,1e-5);
+	EXPECT_NEAR(w[1],-0.88859,1e-5);
+
+	delete opt2;
+	delete updater2;
```
Comment (karlnapf) [inline tests/unit/optimization/StochasticMinimizers_unittest.cc]: hey what about this issue with things not being CSGOBject?
Shall we revert now that plugins are coming?

**Comment by yorkerlin** [comment]
Yes. In fact, we can re-use the code for benchmarking a shogun's algorithm against other implementation(s).

File: tests/unit/optimization/StochasticMinimizers_unittest.cc (hunk 2 above)
Comment (yorkerlin) [inline tests/unit/optimization/StochasticMinimizers_unittest.cc]: I am working on it from this week.

Comment (yorkerlin) [inline tests/unit/optimization/StochasticMinimizers_unittest.cc]: In fact, I will add refactor the framework and add some new algorithms about coordinate descend. 

ref: https://github.com/uclaopt/TMAC
project TMAC is written in C++11 and has template design.
I plan to refactor the existing framework in shogun soon.

**Comment by yorkerlin** [comment]
@karalets 
BTW, when will you release the next version of shogun?
If you plan to release the next version around September, I can refactor and extend shogun's OPT framework.

**Comment by karlnapf** [comment]
After GSoC probably.
For now, please focus on clean-ups and refactorings. If you want to add new things, do that in a feature branch.
I would also prefer some benchmarking or other type of stress testing before we go and add more and more methods.

**Comment by yorkerlin** [comment]
In fact, I am also reading vw's code.
I will Integrate VW into shogun once issue https://github.com/shogun-toolbox/shogun/issues/3316 is solved.

**Comment by vigsterkr** [comment]
@yorkerlin there's already some vw integration within shogun

**Comment by karlnapf** [comment]
Yeah, and that needs update. I talked to John Langford about this a while ago and he said we should ask for help, he will do so

**Comment by yorkerlin** [comment]
https://github.com/shogun-toolbox/shogun/pull/3331 is one PR related to vw modular in shogun.
BTW, the `vw` modular in shogun does not work :(
I will try to fix the issue.

**Comment by yorkerlin** [comment]
I am merging this PR.

**Comment by yorkerlin** [comment]
I am fixing some bugs in `vw` modular since I will use those useful classes in `opt` modular.
`vw` modular will be updated.

**Pull request merged**